code/modules/modular_computers/computers/modular_computer/core.dm

```diff
@@ -23,48 +21,53 @@
 
 // Used to perform preset-specific hardware changes.
 /obj/item/modular_computer/proc/install_default_hardware()
-	return 1
+	var/datum/extension/assembly/assembly = get_extension(src, /datum/extension/assembly)
+	for(var/T in default_hardware)
+		assembly.try_install_component(null, new T(src))
 
 // Used to install preset-specific programs
 /obj/item/modular_computer/proc/install_default_programs()
-	return 1
-
-/obj/item/modular_computer/proc/install_default_programs_by_job(var/mob/living/carbon/human/H)
+	var/mob/living/carbon/human/H = get_holder_of_type(src, /mob)
+	if(!H)
+		return
 	var/datum/job/jb = SSjobs.get_by_title(H.job)
 	if(!jb) return
-	for(var/prog_type in jb.software_on_spawn)
+	var/datum/extension/assembly/modular_computer/assembly = get_extension(src, /datum/extension/assembly)
+	var/obj/item/stock_parts/computer/hard_drive/HDD = assembly.get_component(PART_HDD)
+	if(!HDD)
+		return
+	for(var/prog_type in (jb.software_on_spawn + default_programs))
 		var/datum/computer_file/program/prog_file = prog_type
-		if(initial(prog_file.usage_flags) & hardware_flag)
+		if(initial(prog_file.usage_flags) & assembly.hardware_flag)
 			prog_file = new prog_file
-			hard_drive.store_file(prog_file)
+			HDD.store_file(prog_file)
 
 /obj/item/modular_computer/Initialize()
 	START_PROCESSING(SSobj, src)
 	set_extension(src, /datum/extension/interactive/ntos/device)
+	set_extension(src, /datum/extension/assembly/modular_computer)
 
 	if(stores_pen && ispath(stored_pen))
 		stored_pen = new stored_pen(src)
 
-	install_default_hardware()
-	if(hard_drive)
-		install_default_programs()
-	if(scanner)
-		scanner.do_after_install(null, src)
 	update_icon()
 	update_verbs()
 	update_name()
 	if(enabled_by_default)
 		enable_computer()
+	..()
+	return INITIALIZE_HINT_LATELOAD
```

Comment:
Why lateload it?

Reply:
/obj/item/modular_computer/LateInitialize()
	. = ..()
	install_default_hardware()
	install_default_programs()

Initialization for programs and hardware has to happen later because of race condition reasons.